src/Raven.Server/Documents/CountersStorage.cs

```diff
@@ -493,6 +565,63 @@ private BlittableJsonReaderObject WriteNewCountersDocument(DocumentsOperationCon
             return data;
         }
 
+        internal static (BlittableJsonReaderObject First, BlittableJsonReaderObject Second) SplitCounterDocument(DocumentsOperationContext context, BlittableJsonReaderObject values, BlittableJsonReaderArray dbIds)
+        {
+            // here we rely on the internal sort order of the blittables, because we go through them
+            // in lexical order
+            var fst = CreateHalfDocument(context, values, 0, values.Count / 2, dbIds);
+            var snd = CreateHalfDocument(context, values, values.Count / 2, values.Count, dbIds);
+
+            return (fst, snd);
+        }
+
+        private static BlittableJsonReaderObject CreateHalfDocument(DocumentsOperationContext context, BlittableJsonReaderObject values, int start, int end, BlittableJsonReaderArray dbIds)
+        {
+            BlittableJsonReaderObject.PropertyDetails prop = default;
+            using (var builder = new ManualBlittableJsonDocumentBuilder<UnmanagedWriteBuffer>(context))
+            {
+                context.CachedProperties.NewDocument();
+                builder.Reset(BlittableJsonDocumentBuilder.UsageMode.None);
+                builder.StartWriteObjectDocument();
+
+                builder.StartWriteObject();
+
+                builder.WritePropertyName(DbIds);
+
+                builder.StartWriteArray();
+
+                foreach (LazyStringValue item in dbIds)
+                {
+                    builder.WriteValue(item);
+                }
+
+                builder.WriteArrayEnd();
+
+                builder.WritePropertyName(Values);
+                builder.StartWriteObject();
+
+                for (int i = start; i < end; i++)
+                {
+                    values.GetPropertyByIndex(i, ref prop);
+
+                    builder.WritePropertyName(prop.Name);
+                    if (prop.Value is BlittableJsonReaderObject.RawBlob blob)
+                        builder.WriteRawBlob(blob.Ptr, blob.Length);
+                    else if (prop.Value is LazyStringValue lsv)
+                        builder.WriteValue(lsv);
+                    else
+                        throw new InvalidDataException("Unknown type: " + prop.Token + " when trying to split counter doc");
+                }
+
+                builder.WriteObjectEnd();
+
+                builder.WriteObjectEnd();
+                builder.FinalizeDocument();
+
+                return builder.CreateReader();
+            }
+        }
+
         public void PutCounters(DocumentsOperationContext context, string documentId, string collection, string changeVector,
```

Comment:
This is pretty complex, and I think this requires additional code review (and probably some comments).
Let's sit on this tomorrow.